logging/src/logging.cpp

```diff
@@ -0,0 +1,107 @@
+/*
+    Copyright (c) 2011-2015 Anton Matveenko <antmat@yandex-team.ru>
+    Copyright (c) 2011-2015 Other contributors as noted in the AUTHORS file.
+
+    This file is part of Cocaine.
+
+    Cocaine is free software; you can redistribute it and/or modify
+    it under the terms of the GNU Lesser General Public License as published by
+    the Free Software Foundation; either version 3 of the License, or
+    (at your option) any later version.
+
+    Cocaine is distributed in the hope that it will be useful,
+    but WITHOUT ANY WARRANTY; without even the implied warranty of
+    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+    GNU Lesser General Public License for more details.
+
+    You should have received a copy of the GNU Lesser General Public License
+    along with this program. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include "cocaine/service/logging.hpp"
+
+namespace ph = std::placeholders;
+
+namespace cocaine { namespace service {
+
+const io::basic_dispatch_t&
+logging_t::prototype() {
+    return *this;
+}
+
+logging_t::logging_t(context_t& context, asio::io_service& asio, const std::string& name, const dynamic_t& args)
+{
+    typedef typename io::basic_slot<io::base_log::get>::dispatch_type get_dispatch_type;
+    typedef typename io::basic_slot<io::base_log::get>::tuple_type get_tuple_type;
+    typedef typename io::basic_slot<io::base_log::get>::upstream_type get_upstream_type;
+    auto control = context.get<logging::control_t>();
+    on<io::base_log::get>([=](get_tuple_type&& args, get_upstream_type&& upstream) {
+        return cocaine::tuple::invoke(std::forward(args), [&](std::string name) {
+            auto filter = control.metafilter(name);
+            return boost::make_optional(std::shared_ptr<const dispatch_type>(
+                new named_logging_t(this, std::move(name), std::move(filter))
+            ));
+        });
+    });
+}
+
+named_logging_t::named_logging_t(logging_t* parent_, std::string name_, std::shared_ptr<metafilter_t> filter_) :
+    parent(prent_),
+    name(std::move(name_)),
+    filter(std::move(filter_))
+{
+    on<io::named_log::emit>(std::bind(&named_logging_t::emit, this, ph::_1));
+}
+
+void
+named_logging_t::emit(std::string message, blackhole::attribute::set_t attributes) {
+    if(filter->apply(attributes)) {
+        auto record = log->open_record(level, std::move(attributes));
+
+        if(!record) {
+            return;
+        }
+
+        record.insert(cocaine::logging::keyword::source() = std::move(source));
+        record.message(std::move(message));
+
+        wrapper->push(std::move(record));
+    }
+}
+
+const io::basic_dispatch_t&
+logging_control_t::prototype() const {
+
+}
+
+logging_control_t::logging_control_t(context_t& context, asio::io_service& asio, const std::string& name, const dynamic_t& args) {
+
```

Comment:
Mark with `BOOST_ASSERT(false)` or similar construction to help us see is this part really unimplemented or it is empty just as planned.